ARM64 NEON SIMD huffman encoding

mayeut opened the pull request:
Refer to #44, #47 and #49 for more details.

Timings on the Cavium ThunderX ARM64 board:
master d38b4f21ec5baabb448cd9ffa078fa9150d54af2 (edited 2016-01-17)
```
vgl_6434_0018a.bmp 100 444 10.805636 Megapixels/sec
vgl_6434_0018a.bmp 95 444 13.092066 Megapixels/sec
vgl_6434_0018a.bmp 95 422 16.349982 Megapixels/sec
vgl_6434_0018a.bmp 95 420 20.339437 Megapixels/sec
artificial.ppm 100 444 15.478101 Megapixels/sec
artificial.ppm 95 444 20.059311 Megapixels/sec
artificial.ppm 95 422 26.932211 Megapixels/sec
artificial.ppm 95 420 33.242520 Megapixels/sec
nightshot_iso_100.ppm 100 444 12.438681 Megapixels/sec
nightshot_iso_100.ppm 95 444 19.313331 Megapixels/sec
nightshot_iso_100.ppm 95 422 25.977781 Megapixels/sec
nightshot_iso_100.ppm 95 420 32.086176 Megapixels/sec

vgl_6434_0018a.bmp 10 444 18.169759 Megapixels/sec
vgl_6434_0018a.bmp 50 444 16.888690 Megapixels/sec
vgl_6434_0018a.bmp 95 444 13.120563 Megapixels/sec
artificial.ppm 10 444 21.620149 Megapixels/sec
artificial.ppm 50 444 21.266031 Megapixels/sec
artificial.ppm 95 444 20.155321 Megapixels/sec
nightshot_iso_100.ppm 10 444 21.657411 Megapixels/sec
nightshot_iso_100.ppm 50 444 21.346780 Megapixels/sec
nightshot_iso_100.ppm 95 444 19.325315 Megapixels/sec
```

With this PR  (edited 2016-01-17):
```
vgl_6434_0018a.bmp 100 444 10.178003 Megapixels/sec
vgl_6434_0018a.bmp 95 444 12.460482 Megapixels/sec
vgl_6434_0018a.bmp 95 422 15.655713 Megapixels/sec
vgl_6434_0018a.bmp 95 420 19.546664 Megapixels/sec
artificial.ppm 100 444 13.969191 Megapixels/sec
artificial.ppm 95 444 17.256691 Megapixels/sec
artificial.ppm 95 422 23.585533 Megapixels/sec
artificial.ppm 95 420 29.322864 Megapixels/sec
nightshot_iso_100.ppm 100 444 12.287634 Megapixels/sec
nightshot_iso_100.ppm 95 444 16.835227 Megapixels/sec
nightshot_iso_100.ppm 95 422 23.063786 Megapixels/sec
nightshot_iso_100.ppm 95 420 28.802179 Megapixels/sec

vgl_6434_0018a.bmp 10 444 15.608763 Megapixels/sec
vgl_6434_0018a.bmp 50 444 14.918838 Megapixels/sec
vgl_6434_0018a.bmp 95 444 12.457544 Megapixels/sec
artificial.ppm 10 444 18.014999 Megapixels/sec
artificial.ppm 50 444 17.840014 Megapixels/sec
artificial.ppm 95 444 17.257215 Megapixels/sec
nightshot_iso_100.ppm 10 444 18.039461 Megapixels/sec
nightshot_iso_100.ppm 50 444 17.854830 Megapixels/sec
nightshot_iso_100.ppm 95 444 16.807877 Megapixels/sec
```

With this PR + ZigZag `tbl/tblx` NEON from #47 (edited 2016-01-17):
```
vgl_6434_0018a.bmp 100 444 5.193216 Megapixels/sec
vgl_6434_0018a.bmp 95 444 5.725389 Megapixels/sec
vgl_6434_0018a.bmp 95 422 7.642293 Megapixels/sec
vgl_6434_0018a.bmp 95 420 9.835368 Megapixels/sec
artificial.ppm 100 444 6.056353 Megapixels/sec
artificial.ppm 95 444 6.594384 Megapixels/sec
artificial.ppm 95 422 9.552110 Megapixels/sec
artificial.ppm 95 420 12.386251 Megapixels/sec
nightshot_iso_100.ppm 100 444 5.719931 Megapixels/sec
nightshot_iso_100.ppm 95 444 6.534959 Megapixels/sec
nightshot_iso_100.ppm 95 422 9.447844 Megapixels/sec
nightshot_iso_100.ppm 95 420 12.272391 Megapixels/sec

vgl_6434_0018a.bmp 10 444 6.309603 Megapixels/sec
vgl_6434_0018a.bmp 50 444 6.193853 Megapixels/sec
vgl_6434_0018a.bmp 95 444 5.729896 Megapixels/sec
artificial.ppm 10 444 6.711948 Megapixels/sec
artificial.ppm 50 444 6.688382 Megapixels/sec
artificial.ppm 95 444 6.600798 Megapixels/sec
nightshot_iso_100.ppm 10 444 6.715829 Megapixels/sec
nightshot_iso_100.ppm 50 444 6.695091 Megapixels/sec
nightshot_iso_100.ppm 95 444 6.480968 Megapixels/sec
```

Timings on iPhone 5S:
master d38b4f21ec5baabb448cd9ffa078fa9150d54af2 (edited 2016-01-17)  
```
artificial.ppm 100 444 33.969899 Megapixels/sec
artificial.ppm 95 444 43.914393 Megapixels/sec
artificial.ppm 95 422 65.801160 Megapixels/sec
artificial.ppm 95 420 82.042021 Megapixels/sec
nightshot_iso_100.ppm 100 444 19.755143 Megapixels/sec
nightshot_iso_100.ppm 95 444 41.288920 Megapixels/sec
nightshot_iso_100.ppm 95 422 58.889601 Megapixels/sec
nightshot_iso_100.ppm 95 420 72.865980 Megapixels/sec

artificial.ppm 10 444 56.494088 Megapixels/sec
artificial.ppm 50 444 53.840024 Megapixels/sec
artificial.ppm 95 444 43.914393 Megapixels/sec
nightshot_iso_100.ppm 10 444 58.658252 Megapixels/sec
nightshot_iso_100.ppm 50 444 51.388827 Megapixels/sec
nightshot_iso_100.ppm 95 444 41.288920 Megapixels/sec
```

With this PR (edited 2016-01-17):
```
artificial.ppm 100 444 44.980973 Megapixels/sec
artificial.ppm 95 444 56.373134 Megapixels/sec
artificial.ppm 95 422 85.829465 Megapixels/sec
artificial.ppm 95 420 106.047026 Megapixels/sec
nightshot_iso_100.ppm 100 444 32.613519 Megapixels/sec
nightshot_iso_100.ppm 95 444 53.303646 Megapixels/sec
nightshot_iso_100.ppm 95 422 78.063981 Megapixels/sec
nightshot_iso_100.ppm 95 420 97.723906 Megapixels/sec

artificial.ppm 10 444 66.909933 Megapixels/sec
artificial.ppm 50 444 64.246335 Megapixels/sec
artificial.ppm 95 444 56.373134 Megapixels/sec
nightshot_iso_100.ppm 10 444 69.870525 Megapixels/sec
nightshot_iso_100.ppm 50 444 67.061232 Megapixels/sec
nightshot_iso_100.ppm 95 444 53.303646 Megapixels/sec
```

With this PR + ZigZag `tbl/tblx` NEON from #47 (edited 2016-01-17):
```
artificial.ppm 100 444 47.066281 Megapixels/sec
artificial.ppm 95 444 59.685320 Megapixels/sec
artificial.ppm 95 422 89.740403 Megapixels/sec
artificial.ppm 95 420 111.626894 Megapixels/sec
nightshot_iso_100.ppm 100 444 33.706232 Megapixels/sec
nightshot_iso_100.ppm 95 444 55.344628 Megapixels/sec
nightshot_iso_100.ppm 95 422 82.148109 Megapixels/sec
nightshot_iso_100.ppm 95 420 103.424884 Megapixels/sec

artificial.ppm 10 444 70.874391 Megapixels/sec
artificial.ppm 50 444 67.772399 Megapixels/sec
artificial.ppm 95 444 59.685320 Megapixels/sec
nightshot_iso_100.ppm 10 444 73.730281 Megapixels/sec
nightshot_iso_100.ppm 50 444 70.327086 Megapixels/sec
nightshot_iso_100.ppm 95 444 55.344628 Megapixels/sec
```

dcommander commented:
I can confirm that the ARM64 Huffman SIMD code, as implemented, slows performance on ThunderX by 12% on average for your first implementation and by more than 2x for the zig-zag implementation.

Please do the following:
* Rebase this branch against master, since the rest of the ARM64 NEON routines have been checked in.
* Please do not close this PR.  I prefer that we maintain one discussion thread for each feature instead of you opening a new PR each time you change a commit.  If you change the commits on this branch or add new commits, then it will update this PR, which is what I want.  Let's keep all discussion of ARM64 Huffman SIMD performance here.  When there are multiple PRs for the same feature, it becomes very difficult for me to track the information.
* Please change the title to something that indicates that this PR is solely for ARM64 Huffman encoding.

mayeut commented:
@ssvb, seems you're using the Cavium ThunderX. Not to closer you're results I've just logged off.
Can you tell  me when I can use it ?  thanks

mayeut commented:
> Please do not close this PR. I prefer that we maintain one discussion thread for each feature instead of you opening a new PR each time you change a commit. If you change the commits on this branch or add new commits, then it will update this PR, which is what I want. Let's keep all discussion of ARM64 Huffman SIMD performance here. When there are multiple PRs for the same feature, it becomes very difficult for me to track the information.

I know that new commits are automatically added to the PR, this how I worked on #44. However I didn't find a way to squash all commits from a PR branch (published) into 1 commit (you asked for) in the same branch. That's why I closed  & reopened PRs. Googling around didn't help me with this, maybe I didn't search right.

ssvb commented:
@mayeut Sorry, I was just using ThunderX for some quick experiments with unaligned reads/writes. I have logged off now.

As for squashing commits, you can https://www.kernel.org/pub/software/scm/git/docs/user-manual.html#cleaning-up-history in a branch (using "git reset --hard" and other git commands) and then force push it to github.

mayeut commented:
@ssvb, thanks !

dcommander commented:
To combine commits, use git rebase -i, then force push to the branch.

But I was probably wrong to ask those commits to be combined earlier. I assumed they were in a more final state that I could merge without much modification, but it seems as if all of this stuff is still going to be somewhat in flux until Siarhei and I get ARM64 hardware in house.

I would suggest for now that we do the best we can with what we have available (iPhone and ThunderX), and further testing and optimization can occur later if necessary. I need to keep this moving forward. It seems like the ThunderX is sort of a lowest common denominator, so I suspect that, in most cases, if code performs well on that chip and on the iPhone, it will probably perform well on other chips.

ssvb commented:
The huffman NEON code is a pretty impressive nice optimization.  Just one minor nitpick, unrelated to the correctness and performance. Would it be possible to use human readable aliases for registers via .req/.unreq directives just like the rest of the ARM assembly code does?

And regarding the ThunderX performance problems on huffman code. I have already mentioned this in another pull request discussion. If it turns out that there is no way to have a single universal blended implementation which works optimal on all hardware, then the runtime CPU features detection code can be extended to also identify microarchitecture variants. On ARM the information about the CPU core is pretty much standardized and available in the [MIDR](http://infocenter.arm.com/help/topic/com.arm.doc.ddi0344k/Bgbiddeb.html) system register (it has the vendor id, the core variant id, the major and minor revisions of the core). On Linux, the relevant parts of the MIDR register are exposed in /proc/cpuinfo and can be extracted from it. Here is one example of a script, which attempts to decode this information: https://github.com/dword1511/shell-scripts/blob/master/arminfo
The disadvantage is of course the addition of extra code paths, which need to be exercised during 'make test' (this can be done automatically by using the JSIMD_* environment variable and running multiple passes with different settings).

dcommander commented:
I would be opposed to having multiple SIMD implementations for the same architecture, but I would not be opposed to detecting ThunderX at run time and falling back to the C Huffman implementation when running on that specific chip.  Primarily, this approach would ensure that all code paths get tested irrespective of the presence of ThunderX (the C Huffman encoder will still be used on non-x86 and non-ARM architectures, and of course anyone still using 1.4.x will be using that encoder.)  I unfortunately can't assume that 'make test' will cover all bases.  It never fails that someone comes up with an obscure security issue when fuzzing the code (Mozilla does a lot of that), and it always seems to affect the Huffman code-- however, only a couple of instances of this have affected the encoder (usually they catch issues by creating corrupt JPEG files and trying to decode them.)

From a maintenance point of view, it bothers me to have a code path that is rarely used in real-world situations, because it is often those real-world situations that reveal obscure bugs.  Thus, my opinion on this is:  if you can't (or don't want to) make the Huffman SIMD code run well on ThunderX, let's just use the C Huffman code on that chip.

mayeut commented:
@ssvb agreed for human readable aliases. Since registers were used for different things throughout the code in the initial implementation, it was easier to check for mis reuse of a register. Now that it's cleaner, I can move it further by using human readable aliases.

@dcommander, my knowledge to optimize for a specific micro architecture is limited (in-order, dual-issue, ...) and even with @ssvb input, I'm still missing some bits & pieces. If a "cycle counter" exists for this SOC, like it exists for the [Cortex A8](http://pulsar.webshaker.net/ccc/result.php?lng=fr), I can try do something by relying on this tool. Otherwise, I won't be able to do optimize this for the ThunderX.

If decided not to optimize for ThunderX, I think it's best to wait for you to receive you ARM64 boards in order to choose between linear load + `tbl/tbx` zigzag (which gives the best performances on iPhone 5S) and this implementation.

ssvb commented:
@dcommander 
> I would be opposed to having multiple SIMD implementations for the same architecture,

[...]

> From a maintenance point of view, it bothers me to have a code path that is rarely used in real-world situations, because it is often those real-world situations that reveal obscure bugs.

That's what libjpeg-turbo is already doing on i386. There is a very rarely used MMX code path, which can be nevertheless enforced via the JSIMD_FORCEMMX environment variable for testing.

> I unfortunately can't assume that 'make test' will cover all bases. It never fails that someone comes up with an obscure security issue when fuzzing the code (Mozilla does a lot of that), and it always seems to affect the Huffman code

I guess, it means that the test suite may benefit from some additional improvements. Huffman is indeed a bit too complex to prove that it is correct and provide good testing coverage, so I agree that we need to be very careful with it. But at least the DCT and colorspace conversion code is reasonably simple and has well defined input/output. A lot of code there can only affect pixel data and only a relatively small part of it is responsible for pointer arithmetic.

> Thus, my opinion on this is: if you can't (or don't want to) make the Huffman SIMD code run well on ThunderX, let's just use the C Huffman code on that chip.

Agreed. This is perfectly reasonable. In the first place, we don't even have a NEON implementation which beats the C Huffman code on ThunderX yet.

What I mean is that @mayeut should not worry too much about the lowest common denominator, but focus on implementing whatever works best on iPhone. And then we can see if it is beneficial to have separate implementations (based on benchmark numbers) or a generic implementation is also good enough. In fact, I also consider the generic C implementation of DCT and colorspace conversion as an additional rarely used code path.

ssvb commented:
@mayeut 
> my knowledge to optimize for a specific micro architecture is limited (in-order, dual-issue, ...) and even with @ssvb input, I'm still missing some bits & pieces. If a "cycle counter" exists for this SOC, like it exists for the Cortex A8, I can try do something by relying on this tool. Otherwise, I won't be able to do optimize this for the ThunderX.

I don't have access to any ThunderX documentation either. AFAIK Cortex-A8 is the last properly documented processor with all the cycle timings and pipeline information kindly provided by the vendor. And we also don't have somebody like Agner Fog for ARM, who could provide nice third-party authored optimization manuals. So far I'm using the following method: http://ssvb.github.io/2011/08/03/discovering-instructions-scheduling-secrets.html

For ARM64 it is the following template (this particular example is [estimating a possible pipelining optimization for jsimd_h2v2_downsample_neon](https://github.com/libjpeg-turbo/libjpeg-turbo/pull/49#issuecomment-171899059)):
```assembler
	.cpu cortex-a53+fp+simd
	.file	"1.c"
	.text
	.align	2
	.global	main
	.type	main, %function
main:

#ifndef CPU_CLOCK_FREQUENCY
#error CPU_CLOCK_FREQUENCY must be defined
#endif

#define LOOP_UNROLL_FACTOR   30

        sub         sp, sp, #128
        add         x2, x2, #16
        bic         x2, x2, #15
        add         x3, x2, #64

        ldr         x1, =(CPU_CLOCK_FREQUENCY / LOOP_UNROLL_FACTOR)
        b           1f

    .balign 64
1:
    .rept LOOP_UNROLL_FACTOR

        /* put some code here and use x2 and x3 as scratch buffer pointers */

#ifdef PIPELINED_LOOP
        /* pipelined inner loop of jsimd_h2v2_downsample_neon */
        ld1    {v0.16b}, [x2]
        ld1    {v1.16b}, [x2]
        mov    v4.16b, v16.16b
        uadalp v4.8h, v0.16b
        uadalp v4.8h, v1.16b
        subs   x5, x5, #1
          st1    {v6.8b}, [x3]
        shrn   v6.8b, v4.8h, #2
#else
        /* non-pipelined inner loop of jsimd_h2v2_downsample_neon */
        ld1    {v0.16b}, [x2]
        ld1    {v1.16b}, [x2]
        mov    v4.16b, v16.16b
        uadalp v4.8h, v0.16b
        uadalp v4.8h, v1.16b
        subs   x5, x5, #1
        shrn   v6.8b, v4.8h, #2
          st1    {v6.8b}, [x3]
#endif

    .endr
        subs        x1, x1, #1
        bne         1b

        add         sp, sp, #128
        mov         w0, 0
        ret
```

Results from 2GHz Cavium ThunderX:
<pre>
ssvb@arm64:~$ gcc -DCPU_CLOCK_FREQUENCY=2000000000 -DPIPELINED_LOOP test.S && time ./a.out

real	0m14.995s
user	0m14.990s
sys	0m0.000s

ssvb@arm64:~$ gcc -DCPU_CLOCK_FREQUENCY=2000000000 test.S && time ./a.out

real	0m20.186s
user	0m20.180s
sys	0m0.000s
</pre>

As you can see, we get pretty accurate measurements here (the times are very close to integer numbers). And thus have all reasons to believe that one variant of the loop is taking 15 cycles per iteration (not counting the loop overhead itself) and the other is is taking 20 cycles. Waiting for 15 or 20 seconds is too much, so it is possible to divide the CPU_CLOCK_FREQUENCY constant by 10 and still have reasonable measurements accuracy.

Experimenting with a few simple instruction sequences, one can easily figure out the latencies and throughput of most common instructions.

dcommander commented:
MMX is a special case.  That code was inherited from the original jpeg-6b SIMD implementation, which was developed in 2006.  The MMX code has basically not changed since 2009.  It is a legacy feature that we could probably get rid of if we wanted to, because SSE2-capable 32-bit CPUs have now been in circulation for nearly 15 years.  If MMX created any kind of maintenance problem at all, I would jettison it without a second thought.  But what you're proposing is different.  You're proposing to introduce two separate implementations for the same instruction set, as opposed to two separate instruction sets for the same architecture.  Given that the ARM64 implementation is still not complete (lots of algorithms still missing on the decode side), I don't consider that implementation to be final.  There is going to be continued development on it, and that development becomes more difficult if there are two code paths.

My assumption here is that what we're doing for ThunderX will benefit other in-order cores, but perhaps that assumption is incorrect.  I guess you and I will find out once we get our hardware in house (which board did you end up ordering, BTW?)

It seems like, however, that the instruction ordering stuff may be a good idea regardless.  Perhaps it is only the ld3 and st3 stuff that will prove to be necessary only on ThunderX.  I don't have a lot of insight into that.

dcommander commented:
The other thing I'll say, from a project management point of view, is that I can't delay the release of libjpeg-turbo 1.5 forever.  It's already been delayed much too long, because I was waiting for something compelling that would encourage x86 users to upgrade.  That something was originally supposed to be AVX2, but that project is very slow getting off the ground, unfortunately.  The accelerated SSE2 Huffman routines are providing enough of a speedup to "sell" the release irrespective of the AVX2 enhancements, so AVX2 has been pushed to 1.6.

There is one last feature (#34) that needs to be integrated into 1.5, which will probably take me into February, but I definitely want the beta to go out in February.  So that gives us about a month to figure out the ARM64 issues.

ssvb commented:
@dcommander 
> MMX is a special case. That code was inherited from the original jpeg-6b SIMD implementation, which was developed in 2006. The MMX code has basically not changed since 2009. It is a legacy feature that we could probably get rid of if we wanted to, because SSE2-capable 32-bit CPUs have now been in circulation for nearly 15 years.  If MMX created any kind of maintenance problem at all, I would jettison it without a second thought.

Yes, MMX is old. But then you mention AVX2 in the next message yourself, so even hypothetically dropping MMX some time in the future will not help to have only a single assembly implementation for each CPU architecture.

> But what you're proposing is different. You're proposing to introduce two separate implementations for the same instruction set, as opposed to two separate instruction sets for the same architecture.

Yes, technically ThunderX supports all NEON instructions and can execute them correctly. However some subset of the instructions is implemented so bad, that using it in the optimized code is actually destroying performance. Effectively we have a variant of NEON, with some of the instructions <del>removed</del> blacklisted. This is not very different from having, for example, one x86 processor supporting only SSE2, and another processor supporting SSE2 + SSSE3. If somebody tries to use bad instructions on a processor, then bad things happen. Just different grades of "bad".

> Given that the ARM64 implementation is still not complete (lots of algorithms still missing on the decode side), I don't consider that implementation to be final. There is going to be continued development on it, and that development becomes more difficult if there are two code paths.

When some algorithm is not implemented in assembly at all, it means that this particular algorithm has zero code paths (or a single code path if we also count the C implementation). So nothing gets more difficult. Or probably I don't quite understand what you mean?

You said earlier: <i>"Thus, my opinion on this is: if you can't (or don't want to) make the Huffman SIMD code run well on ThunderX, let's just use the C Huffman code on that chip."</i> And I don't disagree here. I'm actually in favour of just developing/improving a single NEON assembly implementation for each algorithm. Then try to get rid of slow instructions if it is possible and/or easy. If it is difficult or impossible, then just disable this particular function on ThunderX. And if Cavium or somebody else decides to hire somebody to implement something better for ThunderX, then maybe, just maybe, alternative NEON code paths could be re-considered again. Or @mayeut finds a better solution. It is he who is developing this Huffman patch after all.

> My assumption here is that what we're doing for ThunderX will benefit other in-order cores, but perhaps that assumption is incorrect.

Based on the recent findings about the full 128-bit NEON unit width in Cortex-A53, looks like it is the other way around. Assuming comparable instruction latencies, eliminating pipeline stalls via reordering instructions is going to be much more challenging on Cortex-A53 compared to ThunderX. Just because Cortex-A53 has a much higher theoretical peak performance, there is a lot more performance to lose because of pipeline stalls. If one targets Cortex-A53 when optimizing code, the code will be likely already good for ThunderX too (assuming that the code does not use any instruction from the "bad" list). But we will see how it turns out.

> I guess you and I will find out once we get our hardware in house (which board did you end up ordering, BTW?)

It is a [Pine64](https://www.kickstarter.com/projects/pine64/pine-a64-first-15-64-bit-single-board-super-comput) board from the initial pre-production batch. They have been sending board samples to interested software developers for free and I applied for a board. I think that Pine64 has a good potential to become very popular, just because of the price alone.

> There is one last feature (#34) that needs to be integrated into 1.5, which will probably take me into February, but I definitely want the beta to go out in February. So that gives us about a month to figure out the ARM64 issues.

That's good to know. Thanks for sharing the release schedule information. I think that one month is going to be enough.

My own plans used to be (without any particular deadline):
* Eventually get some ARM64 board
* Try to revisit the Linaro's 64-bit NEON code in libjpeg-turbo, because I was never fully satisfied with it.

dcommander commented:
I am not an ARM64 developer, so I am acting solely as a project manager here, and in that role, what I want is for this to perform as well as possible on the most common platforms and to not regress (relative to libjpeg-turbo 1.4.x or libjpeg v6b) on the uncommon ones, and I want to avoid any undue maintenance burdens.  Here's what I would suggest as a way forward:

* If the decompression optimizations for ThunderX are close to being complete, then it makes sense to go ahead and complete those, then I will integrate them into master.  That will at least give you a good foundation upon which to do further optimizations.
* If it appears that optimizing Huffman coding for ThunderX is going to be difficult or that those optimizations will slow down performance on other processors, then I would suggest postponing any further work on that for the moment.
* I will be getting a Nexus 5X probably tonight.  I can grant you both SSH access to it once it's set up.  That should allow you to play with both an A53 and an A57.
* If you feel that the A53 is a better lowest-common-denominator, then I agree that it makes more sense to optimize for that chip and then deal with ThunderX as an afterthought.
* If ThunderX isn't going to be a representative platform (and it seems that it isn't), then I agree with your proposed approach:  if it is easy and not too disruptive for a particular algorithm, then a substitute code path can be used to avoid the slow instructions.  If that isn't easy, then we can just use the C function for that particular algorithm.  It goes without saying that ThunderX isn't going to be used in mobile platforms, so from my point of view, as long as libjpeg-turbo "does no harm" on that platform, it doesn't necessarily have to be optimal.

I don't mind small chunks of code being different for ThunderX and other ARM64 implementations, but I would like to avoid a situation where we have a whole separate assembly file for ThunderX (unless Cavium decides to pay for that work, as you said.)

dcommander commented:
Wasn't able to find the Nexus locally, so I ordered it direct from Google.  Should have it by Monday.  I'll keep you posted.

dcommander commented:
This patch (ARMv8 Huffman encoding) is unfortunately not performing well enough to justify inclusion in its current state.  It needs further optimizations for in-order pipelines.  Here are my results:

ThunderX:  -16 to -8.9% (avg. -13%)

Cortex-A53:  -1.8 to +12% (avg. +2.6%.)  Only vgl_5674_0098.ppm (the most high-frequency among the test images) experiences a significant speedup (5-12%.)  Other images experience no significant speedup.

Cortex-A57:  -3.5 to +21% (avg. +4.8%.)  Only nightshot_iso_100.ppm and vgl_5674_0098.ppm experience a significant speedup (respectively, 2-8% and 7-21%.)  Other images experience no significant speedup.

mayeut commented:
@dcommander,
The last commit should improve speed on lower frequency images. It still does not perform as well a C code on ThunderX but this probably can't be done easily (the NEON store creates a huge bottle neck).
 
Regarding differences between what I'm seeing on iPhone 5S and performances reported on Cortex-A57, I'm wondering if the issue clang vs gcc #3 is the culprit here. Would it be possible to build & test the Android version you're using with clang to check wether this is the issue here ?
Nevertheless, the last commit shall be better anyway.

dcommander commented:
This looks good.

Full-color compression speedups relative to previous commits:
Cortex-A53 (Nexus 5X), Android, 64-bit: 1.1-13% (avg. 6.0%)
Cortex-A57 (Nexus 5X), Android, 64-bit: 0.0-22% (avg. 6.3%)

Question:
Is the code path for TBL_IS_FAST==0 necessary?  It seems like you may have been trying to speed up the algorithm on ThunderX, but I think it's best instead to just disable accelerated Huffman for ThunderX (my version of the patch does that) since the C algorithm is faster.

Also take a look at 46ecffa324be43aab80f6160dc57d98b0a54a704, which introduces a run-time check for slow LD3/ST3 and will currently only disable LD3/ST3 on ThunderX (but this can be easily expanded later on to include other CPUs.)

mayeut commented:
The code path for TBL_IS_FAST can probably be removed. It was indeed an attempt to speedup on ThunderX.
I'm still seeing some huge differences between what you're getting on Android & what I got on iPhone 5S and it would really be a good thing to know wether or not the clang vs gcc issue #3 is the culprit. Commit 499c470b63fb427791b2602fcc0d3f165a81144c which introduces the ARM32 version has only been checked on iPhone on your end as well (according to the commit message). It should probably be checked on some other gcc built platform to check it does not introduce some regression (Nexus 5X can probably run armv7 code I guess).

dcommander commented:
I checked the ARM32 version on my Nexus a while ago.  The Huffman routines speed up performance by 5.0 - 9.4% (avg. 6.6%) on the Cortex-A53 cores and 1.7 - 21% (avg. 6.6%) on the Cortex-A57 cores.

I'm not sure I understand what you mean by "huge differences" between Android and iPhone 5S.  What is the overall speedup you see on the iPhone from the Huffman routines?

#3 refers to a deficiency in the clang optimizer affecting the C version of Huffman encoding.  Obviously we're not using the C version of Huffman encoding, since you replaced it with an assembly version.

dcommander commented:
Actually, upon further testing, the no-tbl code path performs significantly better on Cortex-A53 as well (1.2 - 3.9%, avg. 2.2% faster than the tbl code path.)  @ssvb mentioned that that implementation has a slow tbl instruction.  Also, it demonstrates a very slight (almost immeasurable-- less than 1%) speedup on Cortex-A57.  What does disabling tbl do on your iPhone?

dcommander commented:
Ah, I think I understand what you might be saying.  You're saying that, because you are building your iPhone version of libjpeg-turbo using clang, the relative difference between the C and ARM64 assembly Huffman encoding routines is going to be larger.  And I concur with that assessment.  I think that's why I saw a 30% speedup on my iPhone 4S as well.  I don't use clang for building ARMv7 iOS code, but I use LLVM/GCC, which suffers from a similar optimizer deficiency when compiling jchuff.c.

Unfortunately, the Android NDK on OS X doesn't include a prebuilt version of clang, so I can't readily test it right now.

dcommander commented:
So, in short, what I need right now:
* Overall speedup numbers from iPhone 5S, comparing current git master (46ecffa324be43aab80f6160dc57d98b0a54a704) with 15aaa7f7e21af6066e22ab18fbd02576078351ff (same code w/o ARM64 Huffman.)
* Relative to git master, the speedup or slow-down on the iPhone from #undefining TBL_IS_FAST

mayeut commented:
Numbers for iPhone 5S are in the commit message:
```
ARM64 Huffman SIMD frequency condition

This adds a conditional path depending on high/low-frequency detection.

Comparison against libjpeg-turbo@fe11699

Performances on ThunderX are still not as good as C implementation:
vgl_6434_0018a.bmp 100 444 -7%
vgl_6434_0018a.bmp 95 444 -5%
vgl_6434_0018a.bmp 95 422 -5%
vgl_6434_0018a.bmp 95 420 -5%
artificial.ppm 100 444 -5%
artificial.ppm 95 444 -9%
artificial.ppm 95 422 -7%
artificial.ppm 95 420 -8%
nightshot_iso_100.ppm 100 444 -4%
nightshot_iso_100.ppm 95 444 -14%
nightshot_iso_100.ppm 95 422 -13%
nightshot_iso_100.ppm 95 420 -12%
vgl_6434_0018a.bmp 10 444 -12%
vgl_6434_0018a.bmp 50 444 -11%
artificial.ppm 10 444 -8%
artificial.ppm 50 444 -9%
nightshot_iso_100.ppm 10 444 -7%
nightshot_iso_100.ppm 50 444 -9%

Performances on iPhone5S are much better for low-frequencies:
artificial.ppm 100 444 +39%
artificial.ppm 95 444 +36%
artificial.ppm 95 422 +39%
artificial.ppm 95 420 +39%
artificial.ppm 10 444 +31%
artificial.ppm 50 444 +31%
nightshot_iso_100.ppm 100 444 +68%
nightshot_iso_100.ppm 95 444 +35%
nightshot_iso_100.ppm 95 422 +39%
nightshot_iso_100.ppm 95 420 +45%
nightshot_iso_100.ppm 10 444 +36%
nightshot_iso_100.ppm 50 444 +37%
```

in the form `file quality subsampling performance_ratio`

I will do measurements undefining TBL_IS_FAST once I get home.

mayeut commented:
undefining TBL_IS_FAST on iPhone 5S still results in performance loss (as it did in the first commit, see 1st description of the PR)
Using commit bba7978981e4fc6d0e10f72e95e5e249479ce0e9 and disabling TBL_IS_FAST gives the following against bba7978981e4fc6d0e10f72e95e5e249479ce0e9 alone:
```
artificial.ppm 100 444 -3%
artificial.ppm 95 444 -8%
artificial.ppm 95 422 -5%
artificial.ppm 95 420 -4%
artificial.ppm 10 444 -5%
artificial.ppm 50 444 -4%
nightshot_iso_100.ppm 100 -3%
nightshot_iso_100.ppm 95 444 -3%
nightshot_iso_100.ppm 95 422 -4%
nightshot_iso_100.ppm 95 420 -4%
nightshot_iso_100.ppm 10 444 -5%
nightshot_iso_100.ppm 50 444 -5%
```

If building ARM64 with clang on Android is not possible (that would only confirm a greater speed up), I guess that a check on ARM32 using gcc shall be done to make sure it's faster than C base code which I now very much doubt.

mayeut commented:
You can forget my last comment on ARM32, just saw the change log already mention this.

dcommander commented:
OK, interesting.  Sounds like the best approach, then, is to disable TBL_IS_FAST only on CPUs that benefit from it (Cortex-A53, for instance.)  Thanks for testing.  While I have your attention, can you either post (or send in e-mail if it's a long description) instructions on how to run TJBench on a non-jailbroken iPhone?  It would help me out a lot to be able to use my 6S as an additional data point.

mayeut commented:
Here's how I ran tjbench on a non-jailbroken iPhone:

Create a new iPhone Project based on `Single View Application` template.
Add the following files:
```
rdppm.c
wrppm.c
rdbmp.c
wrbmp.c
tjutil.c
bmp.c
tjbench.c (make a copy of this one, needs modification)
```

modify `main` to `main_tjbench` in `tjbench.c`

I updated the Build Phases as follow:
```
Link Binary With Libraries: libjpegturbo.a
Copy Bundle Resources: artificial.ppm nightspot_iso_100.ppm vgl_6434_0018a.bmp
```

I updated the Build Settings as follow:
```
Architectures: arm64
Enable Bitcode: no
Other Linker Flags: -L/Users/Matt/Dev/libjpeg-turbo/build/.libs (update this for the correct path on your system, will probably be set by Build Phases)
Header Search Paths: /Users/Matt/Dev/libjpeg-turbo/build /Users/Matt/Dev/libjpeg-turbo/libjpeg-turbo (update this for the correct path on your system)
Preprocessor macros: BMP_SUPPORTED PPM_SUPPORTED
```

All the following will be done in the file `ViewController.m`
Declare the entry point for tjbench:
```C
extern int main_tjbench(int argc, char *argv[]);
```
You need to copy the images from the bundle for the rest to work properly, let's define some helper functions in the ViewController implementation for that.
```C
+ (NSString *) applicationDocumentsDirectory
{
	NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
	NSString *basePath = paths.firstObject;
	return basePath;
}

- (void)copyResources {
	NSBundle* mainBundle = [NSBundle mainBundle];
	NSString* oldImagePath = nil;
	NSString* newImagePath = nil;
	
	
	NSFileManager* manager = [NSFileManager defaultManager];
	[manager changeCurrentDirectoryPath:[ViewController applicationDocumentsDirectory]]; // CurrentDirectory is now Documents which is RW
	oldImagePath = [mainBundle pathForResource:@"vgl_6434_0018a" ofType:@"bmp"];
	newImagePath = [[ViewController applicationDocumentsDirectory] stringByAppendingString:@"/vgl_6434_0018a.bmp"];
	[manager copyItemAtPath:oldImagePath toPath:newImagePath error:nil];
	oldImagePath = [mainBundle pathForResource:@"artificial" ofType:@"ppm"];
	newImagePath = [[ViewController applicationDocumentsDirectory] stringByAppendingString:@"/artificial.ppm"];
	[manager copyItemAtPath:oldImagePath toPath:newImagePath error:nil];
	oldImagePath = [mainBundle pathForResource:@"nightshot_iso_100" ofType:@"ppm"];
	newImagePath = [[ViewController applicationDocumentsDirectory] stringByAppendingString:@"/nightshot_iso_100.ppm"];
	[manager copyItemAtPath:oldImagePath toPath:newImagePath error:nil];
	
}
```
Now modify the `viewDidLoad` function to call the tjbench entry point. This is what I used to get the latest figures of this PR on iPhone 5S:
```C
- (void)viewDidLoad {
	[super viewDidLoad];
	char argv0[] = "tjbench";
	char argv1[64] = "nightshot_iso_100.ppm";
	char argv2[64] = "95";
	char argv3[] = "-rgb";
	char argv4[] = "-subsamp";
	char argv5[64] = "444";
	char argv6[] = "-componly";
	char argv7[] = "-nowrite";
	char argv8[64] = "-fastdct";
	
	char* argv[] = {
		argv0,
		argv1,
	  argv2,
		argv3,
		argv4,
		argv5,
		argv6,
		argv7,
		NULL
	};
	char* argvdec[] = {
		argv0,
		argv1,
		argv3,
		argv8,
		argv7,
		NULL
	};
	[self copyResources];
	// Do any additional setup after loading the view, typically from a nib.
	
	/*
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	
	strcpy(argv1, "artificial_444_Q95.jpg");
	strcpy(argv8, "-rgb");
	main_tjbench((sizeof(argvdec)/sizeof(argvdec[0])) - 1, argvdec);
	*/
	
#if 1
	strcpy(argv1, "artificial.ppm");
	strcpy(argv2, "100");
	strcpy(argv5, "444");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	strcpy(argv1, "artificial.ppm");
	strcpy(argv2, "95");
	strcpy(argv5, "444");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	
	strcpy(argv1, "artificial.ppm");
	strcpy(argv2, "95");
	strcpy(argv5, "422");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	strcpy(argv1, "artificial.ppm");
	strcpy(argv2, "95");
	strcpy(argv5, "420");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	
	strcpy(argv1, "artificial.ppm");
	strcpy(argv2, "10");
	strcpy(argv5, "444");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	
	strcpy(argv1, "artificial.ppm");
	strcpy(argv2, "50");
	strcpy(argv5, "444");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	
	strcpy(argv1, "nightshot_iso_100.ppm");
	strcpy(argv2, "100");
	strcpy(argv5, "444");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	strcpy(argv1, "nightshot_iso_100.ppm");
	strcpy(argv2, "95");
	strcpy(argv5, "444");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	
	strcpy(argv1, "nightshot_iso_100.ppm");
	strcpy(argv2, "95");
	strcpy(argv5, "422");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	strcpy(argv1, "nightshot_iso_100.ppm");
	strcpy(argv2, "95");
	strcpy(argv5, "420");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	
	strcpy(argv1, "nightshot_iso_100.ppm");
	strcpy(argv2, "10");
	strcpy(argv5, "444");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
	strcpy(argv1, "nightshot_iso_100.ppm");
	strcpy(argv2, "50");
	strcpy(argv5, "444");
	main_tjbench((sizeof(argv)/sizeof(argv[0])) - 1, argv);
#endif
}
```

As you can see, this is not very clean code but at least, I could run tjbench on the iPhone.
You'll need an Apple Developer account(which is now free) to actually run the app on the iPhone.
